[scroll-animations] Animation-range values provided by JS bindings should be resolved when needed

nmoucht opened the pull request:
#### 406f6c335d9f223f5c6f081a787311ee33e1f426
<pre>
[scroll-animations] Animation-range values provided by JS bindings should be resolved when needed
<a href="https://bugs.webkit.org/show_bug.cgi?id=285946">https://bugs.webkit.org/show_bug.cgi?id=285946</a>
<a href="https://rdar.apple.com/142910399">rdar://142910399</a>

Reviewed by Antoine Quint.

We should resolve animation-range values provided through the JS API when needed to take into account changes
to scroll container size, font size, etc. This is a similar fix to the one made for ViewTimeline insets.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-range-animation.html:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/scroll-timeline-range-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/view-timeline-range-expected.txt:
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::registerNamedScrollTimeline):
(WebCore::AnimationTimelinesController::setTimelineForName):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::intervalForAttachmentRange const):
* Source/WebCore/animation/TimelineRange.cpp:
(WebCore::SingleTimelineRange::lengthForCSSValue):
(WebCore::SingleTimelineRange::parse):
* Source/WebCore/animation/TimelineRange.h:
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::intervalForAttachmentRange const):
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::autoAlignStartTime):
(WebCore::WebAnimation::setBindingsRangeStart):
(WebCore::WebAnimation::setBindingsRangeEnd):
(WebCore::WebAnimation::range):
(WebCore::WebAnimation::progressBasedTimelineSourceDidChangeMetrics):
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::range const): Deleted.

Canonical link: <a href="https://commits.webkit.org/288986@main">https://commits.webkit.org/288986@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/848318c7f240cc42f151d320dd56168f53e6b68e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84935 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4660 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39325 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90087 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35992 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87020 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4750 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12637 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/66101 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23920 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87980 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3623 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77182 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/46368 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3505 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31412 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35065 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32220 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91453 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12274 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8970 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74582 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12504 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72996 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73701 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18085 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16533 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/4308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13241 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12226 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/17678 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12061 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15555 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13806 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -88,7 +88,7 @@ static const std::optional<CSSToLengthConversionData> cssToLengthConversionData(
 Length SingleTimelineRange::lengthForCSSValue(RefPtr<const CSSPrimitiveValue> value, RefPtr<Element> element)
 {
-    if (!value || value->isCalculated() || !element)
+    if (!value || !element)
```
Nice.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -150,28 +150,26 @@ SingleTimelineRange SingleTimelineRange::parse(TimelineRangeValue&& value, RefPt
             if (rangeOffset.offset)
-                return range(CSSValuePair::createNoncoalescing(*consumedRangeName, *rangeOffset.offset->toCSSValue()), type, nullptr, element);
-            return range(*consumedRangeName, type, nullptr, element);
+                return RefPtr<CSSValue> { CSSValuePair::createNoncoalescing(*consumedRangeName, *rangeOffset.offset->toCSSValue()) };
```
Is the `RefPtr<CSSValue>` absolutely necessary here? Could you not specify it as the lambda's return type?

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -150,28 +150,26 @@ SingleTimelineRange SingleTimelineRange::parse(TimelineRangeValue&& value, RefPt
-            return range(*offset->toCSSValue(), type, nullptr, element);
-        return SingleTimelineRange { };
+            return offset->toCSSValue();
+        return RefPtr<CSSValue> { nullptr };
```
Surely just `return nullptr` would work here with the type specified as the lambda's return vale.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/WebAnimation.h:
```diff
@@ -255,6 +255,9 @@ class WebAnimation : public RefCounted<WebAnimation>, public EventTarget, public
     uint64_t m_globalPosition { 0 };
+    RefPtr<CSSValue> m_specifiedRangeStart;
+    RefPtr<CSSValue> m_specifiedRangeEnd;
+
```
I think these values should be higher up with other `RefPtr<>` types for padding purposes. There definitely should not be a new empty line here.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/WebAnimation.cpp:
```diff
@@ -1923,11 +1923,23 @@ void WebAnimation::setRange(TimelineRange range)
+    if (m_specifiedRangeStart && keyframeEffect)
+        m_timelineRange.start = SingleTimelineRange::range(*m_specifiedRangeStart, SingleTimelineRange::Type::Start, nullptr, keyframeEffect->target());
+    if (m_specifiedRangeEnd && keyframeEffect)
+        m_timelineRange.end = SingleTimelineRange::range(*m_specifiedRangeEnd, SingleTimelineRange::Type::End, nullptr, keyframeEffect->target());
```
You should wrap this all up in a `if (RefPtr keyframeEffect = dynamicDowncast<KeyframeEffect>(m_effect))` statement.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/WebAnimation.cpp:
```diff
@@ -1890,10 +1890,10 @@ void WebAnimation::setBindingsRangeStart(TimelineRangeValue&& rangeStartValue)
 
     auto rangeStart = SingleTimelineRange::parse(WTFMove(rangeStartValue), keyframeEffect->target(), SingleTimelineRange::Type::Start);
-    if (rangeStart == m_timelineRange.start)
+    if (m_specifiedRangeEnd == rangeStart)
```
Is that check correct? Did you mean to compare `m_specifiedRangeStart` here rather?

nmoucht replied:
Fixed.

nmoucht committed:
[scroll-animations] Animation-range values provided by JS bindings should be resolved when needed
https://bugs.webkit.org/show_bug.cgi?id=285946
rdar://142910399

Reviewed by Antoine Quint.

We should resolve animation-range values provided through the JS API when needed to take into account changes
to scroll container size, font size, etc. This is a similar fix to the one made for ViewTimeline insets.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-range-animation.html:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/scroll-timeline-range-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/view-timeline-range-expected.txt:
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::registerNamedScrollTimeline):
(WebCore::AnimationTimelinesController::setTimelineForName):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::intervalForAttachmentRange const):
* Source/WebCore/animation/TimelineRange.cpp:
(WebCore::SingleTimelineRange::lengthForCSSValue):
(WebCore::SingleTimelineRange::parse):
* Source/WebCore/animation/TimelineRange.h:
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::intervalForAttachmentRange const):
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::autoAlignStartTime):
(WebCore::WebAnimation::setBindingsRangeStart):
(WebCore::WebAnimation::setBindingsRangeEnd):
(WebCore::WebAnimation::range):
(WebCore::WebAnimation::progressBasedTimelineSourceDidChangeMetrics):
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::range const): Deleted.

Canonical link: https://commits.webkit.org/288986@main
